File: python/Unidad_1/ej4.py
# Imports
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits import mplot3d

# Config plot
def config_plt():
    fig = plt.figure(figsize=(6,4))
    ax3d = plt.axes(projection="3d")
    return fig, ax3d

# Ejercicios
def ejer_1():
    xdata = np.arange(start = -5, stop = 5, step = 0.01)
    ydata = np.arange(start = -5, stop = 5, step = 0.01)
    X,Y = np.meshgrid(xdata,ydata)
    Z = pow(base = X, exp = 2) + pow(base = Y, exp = 2)
    return X, Y, Z

def ejer_2():
    xdata = np.arange(start = -2, stop = 2, step = 0.01)
    ydata = np.arange(start = -2, stop = 2, step = 0.01)
    X,Y = np.meshgrid(xdata,ydata)
    Z = X * np.exp(-pow(base = X, exp = 2) - pow(base = Y, exp = 2))
    return X, Y, Z

def ejer_3():
    xdata = np.arange(start = -3, stop = 3, step = 0.01)
    ydata = np.arange(start = -3, stop = 3, step = 0.01)
    X,Y = np.meshgrid(xdata,ydata)
    Z = (X * Y * (pow(base = X, exp = 2) - pow(base = Y, exp = 2))) / (pow(base = X, exp = 2) + pow(base = Y, exp = 2))
    return X, Y, Z

def ejer_4():
    xdata = np.arange(start = -8, stop = 8, step = 0.1)
    ydata = np.arange(start = -8, stop = 8, step = 0.1)
    X,Y = np.meshgrid(xdata,ydata)
    module = np.sqrt((pow(base = X, exp = 2) + pow(base = Y, exp = 2)))
    Z = np.sin(module) / module
    return X, Y, Z


# Lista de funciones para recorrer en el for
fList = [ejer_1(), ejer_2(), ejer_3(), ejer_4()]
for n, fun in enumerate(fList):
    X, Y, Z = fList[n]
    # Ploteo
    fig, ax3d = config_plt()
    plt.figure(n + 1)
    ax3d = plt.axes(projection='3d')
    surf=ax3d.plot_surface(X, Y, Z, rstride=7, cstride=7, cmap="gray")
    fig.colorbar(surf, ax=ax3d)
    ax3d.set_title(f'Plot de superficie / Ejercicio {n+1}')
    ax3d.set_xlabel('X')
    ax3d.set_ylabel('Y')
    ax3d.set_zlabel('Z')

# Curva de nivel de una función elegida
# X, Y, Z = ejer_4()
# TODO: 15 curvas de nivel

# plt.savefig("Customized Surface Plot.png")

# Muestro todos los plots
plt.show()

File: matlab/Unidad_1/ej4.m
% 4) Grafique las siguientes superficies utilizando los comandos surf y mesh.
% Use diferentes mapas de colores en cada superficie. 

%% 4.1) z = x^2 + y^2 para x,y entre [-5;5]
figure()
[x_1, y_1] = meshgrid(-5:.5:5);
z_1 = x_1.^2 + y_1.^2;

subplot(2,4,1);
surf(z_1);
title('Surf z = x^2 + y^2');

subplot(2,4,5);
mesh(z_1);
title('Mesh z = x^2 + y^2');


%% 4.2) z = x * e^-(x^2 + y^2) para x,y entre [-2;2]

[x_2, y_2] = meshgrid(-2:.2:2);
z_2 = x_2 .* exp(-(x_2.^2 + y_2.^2));

subplot(2,4,2);
surf(z_2);
title('Surf z = x * e^-(x^2 + y^2)');

subplot(2,4,6);
mesh(z_2);
title('Mesh z = x * e^-(x^2 + y^2)');

%% 4.3) z = (x*y*(x^2-y^2))/(x^2+y^2) para x,y entre [-3;3]

[x_3, y_3] = meshgrid(-3:.3:3);
z_3 = (x_3.*y_3.*(x_3.^2-y_3.^2))./(x_3.^2+y_3.^2);

subplot(2,4,3);
surf(z_3);
title('Surf z = (x*y*(x^2-y^2))/(x^2+y^2)');

subplot(2,4,7);
mesh(z_3);
title('Mesh z = (x*y*(x^2-y^2))/(x^2+y^2)');

%% 4.4) z = sen(sqrt(x^2+y^2))/sqrt(x^2+y^2) para x,y  [-8;8]

[x_4, y_4] = meshgrid(-8:.5:8);
raiz = sqrt(x_4.^2+y_4.^2);
z_4 = sin(raiz)./raiz;

subplot(2,4,4);
surf(z_4);
title('Surf z = sen(sqrt(x^2+y^2))/sqrt(x^2+y^2)');

subplot(2,4,8);
mesh(z_4);
title('Mesh z = sen(sqrt(x^2+y^2))/sqrt(x^2+y^2)');

%% Curvas de  contorno de la funcion 4.4)
figure();
contour(z_3,15);